platforms/core-runtime/build-operations-trace/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java

```diff
@@ -615,4 +660,95 @@ public void finished(BuildOperationDescriptor buildOperation, OperationFinishEve
             }
         }
     }
+
+    interface TraceWriter {
+
+        /**
+         * Write a serialized operation to the log file.
+         */
+        void write(SerializedOperation serializedOperation);
+
+        /**
+         * This method must be called after all write operations have been completed.
+         */
+        void complete(boolean writeTree);
+
+    }
+
+    /**
+     * A {@link TraceWriter} that offloads all writing operations to a separate thread.
```

Comment:
That does look awfully similar to what I wrote here. There are some behavior differences though:
In my implementation, after a single exception is thrown, all future submissions fail. A failure of one async action means the system is in a bad state and future actions should not be executed.
 
In the existing implementation, only the first event to be submitted after an error is rejected. Afterwards, IO actions can be submitted again. This doesn't really seem right.

Some comments on the existing implementation

1. The There is only one `AsyncIoScopeFactory` instance per build, and therefore a single thread. Do we want all users of this service to funnel through a single thread? I would imagine we would want more than once instance of this `AsyncIoScopeFactory` for multiple users

3. The following code looks a bit concerning: 
```
        override fun io(action: () -> Unit) {
            // Fail fast
            checkForFailure()
            pending = submit(action)
        }

        override fun close() {
            pending?.get(settings.ioActionTimeoutMs, TimeUnit.MILLISECONDS)
            checkForFailure()
        }
```

Imagine two threads call `io(Action)`. The first one calls `submit` but does not set the value of `pending` yet. Then, the next thread calls `submit` and sets `pending`. The first thread then overwrites `pending`. That means `pending` is set to the earlier future. 

Afterwards, calling `close` would only wait for the _first_ IO event to finish but would not wait for the second IO event. Instead, calling `shutdown` would wait for all events to finish and would mean we do not need to track `pending` at all. 

Assuming we can come up with a single implementation that is suitable for this use case and the existing use case, I think it is a good idea to have a single implementation of this pattern.

Is there an easy way to use the existing implementation from `BuildOperationTrace`? `AsyncIOScopeFactory` lives in `:kotlin-dsl`

Reply:
> In the existing implementation, only the first event to be submitted after an error is rejected.

The rationale was that exception handling and recovery was best left to the consumer of the API since IO actions are abstract (`() -> Unit`).

> There is only one AsyncIoScopeFactory instance per build, and therefore a single thread. Do we want all users of this service to funnel through a single thread?

This can be fine tuned.

> Imagine two threads call `io(Action)`.

Each `IOScope` can only be used from a single thread at a time. ([IOScope documentation](https://github.com/gradle/gradle/blob/35cf1f2eaa35c25c7c11474925a0ce6917e9282c/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/concurrent/IO.kt#L61))

> Assuming we can come up with a single implementation that is suitable for this use case and the existing use case, I think it is a good idea to have a single implementation of this pattern.

My thoughts exactly 👍 

> Is there an easy way to use the existing implementation from BuildOperationTrace? AsyncIOScopeFactory lives in :kotlin-dsl

It would have to be extracted and moved to some module under `core-runtime` first.